ppa/archive/eebo_tcp.py

```diff
@@ -28,26 +28,100 @@ class Page(xmlmap.XmlObject):
     # use following axis to find all text nodes following this page beginning
     text_contents = xmlmap.StringListField("following::text()")
 
+    # notes on or after this page, for managing footnote text
+    notes = xmlmap.NodeListField("following::NOTE", xmlmap.XmlObject)
+
     def __repr__(self):
         return f"<Page {self.number or '-'} ({self.section_type})>"
 
+    # footnote indicators
+    note_marks = ["*", "†", "‡", "§"]
+    num_note_marks = len(note_marks)
+
+    def get_note_mark(self, i):
+        """Generate a note marker based on note index and :attr:`note_marks`;
+        symbols are used in order and then doubled, tripled, etc as needed."""
+
+        # use modulo to map to the list of available marks
+        mark_index = i % self.num_note_marks
+        # use division to determine how many times to repeat the mark
+        repeat = int(i / self.num_note_marks) + 1
+        return self.note_marks[mark_index] * repeat
+
+    def text_inside_note(self, text):
+        """check if a text element occurs within a NOTE element; if so,
+        return the note element"""
+
+        # bail out if there are no notes following this page
+        if not self.notes:
+            return None
+
+        # check if this text is directly inside a note tag
+        within_note = None
+        # check if this is normal text directly inside a note tag
+        parent = text.getparent()
+        if parent.tag == "NOTE" and text.is_text:
+            within_note = parent
+        # otherwise, check if text is nested somewhere under a note tag
+        else:
+            note_ancestors = parent.xpath("ancestor::NOTE[1]")
+            within_note = note_ancestors[0] if note_ancestors else None
+
+        return within_note
+
+    def note_index(self, note):
+        # given a note element, determine the 0-based index on this page
+
+        # use a for loop so we can bail out once we get a match
+        for i, n in enumerate(self.notes):
+            if n.node == note:
+                return i
+
+        # in normal use the note should be found; raise an exception
+        # if it is not so this will fail loudly
+        raise ValueError
+
     def page_contents(self):
         """generator of text strings between this page beginning tag and
         the next one"""
 
         # strictly speaking we are returning lxml "smart strings"
         # (lxml.etree._ElementUnicodeResult)
 
+        # collect any notes and include after main page text contents
+        notes = []
```

Comment:
is there a reason this doesn't correspond to `self.notes`?

Reply:
hmm, I should probably rename those to clarify/differentiate... `self.notes` is the xmlobject reference to all notes on or after the current page; this local `notes` variable is a list of plain text notes contents to be yielded as page content after the main text of the page. Any thoughts on names?